File: api/src/handlers/programmer.rs
use crate::models::fixtures::*;
use crate::models::programmer::*;
use crate::RuntimeApi;
use futures::stream::{Stream, StreamExt};
use mizer_command_executor::*;
use mizer_fixtures::manager::FixtureManager;
use mizer_fixtures::programmer::ProgrammerView;
use mizer_fixtures::GroupId;
use mizer_sequencer::Sequencer;
use std::ops::Deref;

#[derive(Clone)]
pub struct ProgrammerHandler<R> {
    fixture_manager: FixtureManager,
    runtime: R,
}

impl<R: RuntimeApi> ProgrammerHandler<R> {
    pub fn new(fixture_manager: FixtureManager, _: Sequencer, runtime: R) -> Self {
        Self {
            fixture_manager,
            runtime,
        }
    }

    #[tracing::instrument(skip(self))]
    pub fn state_stream(&self) -> impl Stream<Item = ProgrammerState> + 'static {
        let programmer = self.fixture_manager.get_programmer();
        programmer.bus().map(ProgrammerState::from)
    }

    #[tracing::instrument(skip(self))]
    pub fn write_control(&self, request: WriteControlRequest) {
        let mut programmer = self.fixture_manager.get_programmer();
        let control = request.as_controls();
        programmer.write_control(control);
    }

    #[tracing::instrument(skip(self))]
    pub fn select_fixtures(&self, fixture_ids: Vec<FixtureId>) {
        log::debug!("select_fixtures {:?}", fixture_ids);
        let mut programmer = self.fixture_manager.get_programmer();
        programmer.select_fixtures(fixture_ids.into_iter().map(|id| id.into()).collect());
    }

    #[tracing::instrument(skip(self))]
    pub fn unselect_fixtures(&self, fixture_ids: Vec<FixtureId>) {
        log::debug!("unselect_fixtures {:?}", fixture_ids);
        let mut programmer = self.fixture_manager.get_programmer();
        programmer.unselect_fixtures(fixture_ids.into_iter().map(|id| id.into()).collect());
    }

    #[tracing::instrument(skip(self))]
    pub fn select_group(&self, group_id: u32) {
        log::debug!("select_group {:?}", group_id);
        if let Some(group) = self.fixture_manager.get_group(GroupId(group_id)) {
            let mut programmer = self.fixture_manager.get_programmer();
            programmer.select_group(&group);
        }
    }

    #[tracing::instrument(skip(self))]
    pub fn clear(&self) {
        let mut programmer = self.fixture_manager.get_programmer();
        programmer.clear();
    }

    #[tracing::instrument(skip(self))]
    pub fn highlight(&self, highlight: bool) {
        let mut programmer = self.fixture_manager.get_programmer();
        programmer.set_highlight(highlight);
    }

    #[tracing::instrument(skip(self))]
    pub fn store(&self, sequence_id: u32, store_mode: store_request::Mode, cue_id: Option<u32>) {
        let (controls, effects) = {
            let programmer = self.fixture_manager.get_programmer();
            let controls = programmer.get_controls();
            let effects = programmer.active_effects().cloned().collect::<Vec<_>>();

            (controls, effects)
        };
        self.runtime
            .run_command(StoreProgrammerInSequenceCommand {
                sequence_id,
                cue_id,
                controls,
                store_mode: store_mode.into(),
                effects,
            })
            .unwrap();
    }

    #[tracing::instrument(skip(self))]
    pub fn get_presets(&self) -> Presets {
        Presets {
            intensities: self
                .fixture_manager
                .presets
                .intensity_presets()
                .into_iter()
                .map(Preset::from)
                .collect(),
            shutters: self
                .fixture_manager
                .presets
                .shutter_presets()
                .into_iter()
                .map(Preset::from)
                .collect(),
            colors: self
                .fixture_manager
                .presets
                .color_presets()
                .into_iter()
                .map(Preset::from)
                .collect(),
            positions: self
                .fixture_manager
                .presets
                .position_presets()
                .into_iter()
                .map(Preset::from)
                .collect(),
            ..Default::default()
        }
    }

    #[tracing::instrument(skip(self))]
    pub fn call_preset(&self, preset_id: PresetId) {
        log::debug!("call_preset {:?}", preset_id);
        let mut programmer = self.fixture_manager.get_programmer();
        programmer.call_preset(&self.fixture_manager.presets, preset_id.into());
    }

    #[tracing::instrument(skip(self))]
    pub fn call_effect(&self, effect_id: u32) {
        log::debug!("call_effect {:?}", effect_id);
        let mut programmer = self.fixture_manager.get_programmer();
        programmer.call_effect(effect_id);
    }

    #[tracing::instrument(skip(self))]
    pub fn get_groups(&self) -> Groups {
        Groups {
            groups: self
                .fixture_manager
                .get_groups()
                .into_iter()
                .map(|group| group.deref().clone().into())
                .collect(),
            ..Default::default()
        }
    }

    #[tracing::instrument(skip(self))]
    pub fn add_group(&self, name: String) -> Group {
        let group = self.runtime.run_command(AddGroupCommand { name }).unwrap();

        group.into()
    }

    #[tracing::instrument(skip(self))]
    pub fn assign_fixtures_to_group(&self, group_id: u32, fixture_ids: Vec<FixtureId>) {
        let group_id = GroupId(group_id);
        let fixture_ids = fixture_ids
            .into_iter()
            .map(|id| id.into())
            .collect::<Vec<_>>();
        self.runtime
            .run_command(AssignFixturesToGroupCommand {
                group_id,
                fixture_ids,
            })
            .unwrap();
    }

    #[tracing::instrument(skip(self))]
    pub fn assign_fixture_selection_to_group(&self, group_id: u32) {
        let group_id = GroupId(group_id);
        let fixture_ids = {
            let programmer = self.fixture_manager.get_programmer();
            let state = programmer.view().read();
            state.all_fixtures()
        };

        self.runtime
            .run_command(AssignFixturesToGroupCommand {
                group_id,
                fixture_ids,
            })
            .unwrap();
    }

    #[tracing::instrument(skip(self))]
    pub fn programmer_view(&self) -> ProgrammerView {
        self.fixture_manager.get_programmer().view()
    }

    #[tracing::instrument(skip(self))]
    pub fn update_block_size(&self, block_size: usize) {
        self.fixture_manager
            .get_programmer()
            .set_block_size(block_size);
    }

    #[tracing::instrument(skip(self))]
    pub fn update_groups(&self, groups: usize) {
        self.fixture_manager.get_programmer().set_groups(groups);
    }

    #[tracing::instrument(skip(self))]
    pub fn update_wings(&self, wings: usize) {
        self.fixture_manager.get_programmer().set_wings(wings);
    }

    #[tracing::instrument(skip(self))]
    pub fn next(&self) {
        self.fixture_manager.get_programmer().next();
    }

    #[tracing::instrument(skip(self))]
    pub fn prev(&self) {
        self.fixture_manager.get_programmer().prev();
    }

    #[tracing::instrument(skip(self))]
    pub fn set(&self) {
        self.fixture_manager.get_programmer().set();
    }

    #[tracing::instrument(skip(self))]
    pub fn shuffle(&self) {
        self.fixture_manager.get_programmer().shuffle();
    }

    #[tracing::instrument(skip(self))]
    pub fn write_effect_rate(&self, request: WriteEffectRateRequest) {
        self.fixture_manager
            .get_programmer()
            .write_rate(request.effect_id, request.effect_rate);
    }

    #[tracing::instrument(skip(self))]
    pub fn write_effect_offset(&self, request: WriteEffectOffsetRequest) {
        self.fixture_manager
            .get_programmer()
            .write_offset(request.effect_id, request.effect_offset);
    }
}

File: crates/ui/lib/api/mobile/programmer.dart
import 'package:grpc/grpc.dart';
import 'package:mizer/api/plugin/ffi/programmer.dart';
import 'package:mizer/protos/fixtures.pb.dart';
import 'package:mizer/protos/programmer.pbgrpc.dart';

import '../contracts/programmer.dart';

class ProgrammerMobileApi extends ProgrammerApi {
  final ProgrammerApiClient _client;

  ProgrammerMobileApi(ClientChannel channel) : _client = ProgrammerApiClient(channel);

  @override
  Future<Group> addGroup(String name) {
    // TODO: implement addGroup
    throw UnimplementedError();
  }

  @override
  Future<void> assignFixtureSelectionToGroup(Group group) {
    // TODO: implement assignFixtureSelectionToGroup
    throw UnimplementedError();
  }

  @override
  Future<void> assignFixturesToGroup(List<FixtureId> fixtures, Group group) {
    // TODO: implement assignFixturesToGroup
    throw UnimplementedError();
  }

  @override
  Future<void> callEffect(int id) {
    // TODO: implement callEffect
    throw UnimplementedError();
  }

  @override
  Future<void> callPreset(PresetId id) {
    // TODO: implement callPreset
    throw UnimplementedError();
  }

  @override
  Future<void> clear() async {
    await _client.clear(ClearRequest());
  }

  @override
  Future<void> deleteGroup(int id) {
    // TODO: implement deleteGroup
    throw UnimplementedError();
  }

  @override
  Future<void> deletePreset(PresetId id) {
    // TODO: implement deletePreset
    throw UnimplementedError();
  }

  @override
  Future<Groups> getGroups() {
    // TODO: implement getGroups
    throw UnimplementedError();
  }

  @override
  Future<Presets> getPresets() {
    // TODO: implement getPresets
    throw UnimplementedError();
  }

  @override
  Future<ProgrammerStatePointer?> getProgrammerPointer() {
    // TODO: implement getProgrammerPointer
    throw UnimplementedError();
  }

  @override
  Future<void> highlight(bool highlight) async {
    await _client.highlight(HighlightRequest()..highlight = highlight);
  }

  @override
  Future<void> next() {
    // TODO: implement next
    throw UnimplementedError();
  }

  @override
  Stream<ProgrammerState> observe() {
    return this._client.subscribeToProgrammer(SubscribeProgrammerRequest());
  }

  @override
  Future<void> prev() {
    // TODO: implement prev
    throw UnimplementedError();
  }

  @override
  Future<void> renameGroup(int id, String name) {
    // TODO: implement renameGroup
    throw UnimplementedError();
  }

  @override
  Future<void> renamePreset(PresetId id, String name) {
    // TODO: implement renamePreset
    throw UnimplementedError();
  }

  @override
  Future<void> selectFixtures(List<FixtureId> fixtureIds) async {
    await _client.selectFixtures(SelectFixturesRequest()..fixtures.addAll(fixtureIds));
  }

  @override
  Future<void> selectGroup(int id) {
    // TODO: implement selectGroup
    throw UnimplementedError();
  }

  @override
  Future<void> set() {
    // TODO: implement set
    throw UnimplementedError();
  }

  @override
  Future<void> shuffle() {
    // TODO: implement shuffle
    throw UnimplementedError();
  }

  @override
  Future<void> store(int sequenceId, StoreRequest_Mode storeMode, {int? cueId}) {
    // TODO: implement store
    throw UnimplementedError();
  }

  @override
  Future<void> storePreset(StorePresetRequest request) {
    // TODO: implement storePreset
    throw UnimplementedError();
  }

  @override
  Future<void> unselectFixtures(List<FixtureId> fixtureIds) async {
    await _client.unselectFixtures(UnselectFixturesRequest()..fixtures.addAll(fixtureIds));
  }

  @override
  Future<void> updateBlockSize(int blockSize) {
    // TODO: implement updateBlockSize
    throw UnimplementedError();
  }

  @override
  Future<void> updateGroups(int groups) {
    // TODO: implement updateGroups
    throw UnimplementedError();
  }

  @override
  Future<void> updateWings(int wings) {
    // TODO: implement updateWings
    throw UnimplementedError();
  }

  @override
  Future<void> writeControl(WriteControlRequest request) {
    // TODO: implement writeControl
    throw UnimplementedError();
  }

  @override
  Future<void> writeEffectOffset(int effectId, double? effectOffset) {
    // TODO: implement writeEffectOffset
    throw UnimplementedError();
  }

  @override
  Future<void> writeEffectRate(int effectId, double effectRate) {
    // TODO: implement writeEffectRate
    throw UnimplementedError();
  }
}
